src/windows/list.cpp

```diff
@@ -0,0 +1,687 @@
+
+    GetMountpoints(deviceNumber, &device->mountpoints);
+
+    hPhysical = CreateFileA(
+      device->raw.c_str(), handleOpenFlags,
+      FILE_SHARE_READ | FILE_SHARE_WRITE, NULL,
+      OPEN_EXISTING, FILE_ATTRIBUTE_NORMAL, NULL);
+
+    if (hPhysical == INVALID_HANDLE_VALUE) {
+      errorCode = GetLastError();
+      device->error = "Couldn't open handle to physical device: Error " +
+        std::to_string(errorCode);
+      result = false;
+      break;
+    }
+
+    if (!GetDeviceSize(hPhysical, device)) {
+      errorCode = GetLastError();
+      device->error = "Couldn't get disk geometry: Error " +
+        std::to_string(errorCode);
+      result = false;
+      break;
+    }
+
+    if (!GetAdapterInfo(hPhysical, device)) {
+      errorCode = GetLastError();
+      device->error = "Couldn't get device adapter descriptor: Error " +
+        std::to_string(errorCode);
+      result = false;
+      break;
+    }
+
+    // NOTE: No need to fail over this one,
+    // as we can safely default to a 512B block size
+    if (!GetDeviceBlockSize(hPhysical, device)) {
+      errorCode = GetLastError();
+      printf("[INFO] Couldn't get block size: Error %u\n", errorCode);
+    }
+
+    BOOL isWritable = DeviceIoControl(
+      hPhysical, IOCTL_DISK_IS_WRITABLE, NULL, 0,
+      NULL, 0, &size, NULL);
+
+    device->isReadOnly = !isWritable;
+  }
+
+  if (hDevice != INVALID_HANDLE_VALUE) {
+    CloseHandle(hDevice);
+  }
+
+  if (hPhysical != INVALID_HANDLE_VALUE) {
+    CloseHandle(hPhysical);
+  }
+
+  free(deviceDetailData);
+
+  return result;
+}
+
+std::vector<DeviceDescriptor> ListStorageDevices() {
+  HDEVINFO hDeviceInfo = NULL;
+  SP_DEVINFO_DATA deviceInfoData;
+  std::vector<DeviceDescriptor> deviceList;
+
+  DWORD i;
+  char *enumeratorName;
+  DeviceDescriptor device;
+
+  hDeviceInfo = SetupDiGetClassDevsA(
+    &GUID_DEVICE_INTERFACE_DISK, NULL, NULL,
+    DIGCF_PRESENT | DIGCF_DEVICEINTERFACE);
+
+  if (hDeviceInfo == INVALID_HANDLE_VALUE) {
+    printf("[ERROR] Invalid DeviceInfo handle\n");
+    return deviceList;
+  }
+
+  deviceInfoData.cbSize = sizeof(SP_DEVINFO_DATA);
+
+  for (i = 0; SetupDiEnumDeviceInfo(hDeviceInfo, i, &deviceInfoData); i++) {
+    enumeratorName = GetEnumeratorName(hDeviceInfo, deviceInfoData);
+
+    // If it failed to get the SPDRP_ENUMERATOR_NAME, skip it
+    if (enumeratorName == NULL) {
+      continue;
+    }
+
+    device = DeviceDescriptor();
+
+    device.enumerator = std::string(enumeratorName);
+    device.description = GetFriendlyName(hDeviceInfo, deviceInfoData);
+    device.isRemovable = IsRemovableDevice(hDeviceInfo, deviceInfoData);
+    device.isVirtual = IsVirtualHardDrive(hDeviceInfo, deviceInfoData);
+    device.isSCSI = IsSCSIDevice(enumeratorName);
+    device.isUSB = IsUSBDevice(enumeratorName);
+    device.isCard = device.enumerator == "SD";
+    // device.isSystem = IsSystemDevice(hDeviceInfo, deviceInfoData);
+    device.isSystem = device.isSCSI && !device.isRemovable;
```

Comment:
Huh? What about computers still running off IDE drives? Or NVMe SSDs?

Reply:
They all get enumerated under the SCSI storage driver in Windows, no matter the bus type (which is why bus type & enumerator are two different things); What this essentially checks is whether it's connected to an internal bus.

Reply:
Hmmm, so does that mean `.isSCSI` doesn't actually mean what people might expect it to mean? :-/

Reply:
I don't think so – SCSI being a collection of protocols / bus types, and Windows enumerating them under the SCSI storage devices yields what you would expect it to on this platform.